soccer/gameplay/plays/Attack_Goal/adaptive_attack.py

```diff
@@ -0,0 +1,292 @@
+        if (self.shot_chance > AdaptiveAttack.DRIBBLE_TO_SHOOT_CUTOFF):
+            print("Pass : " + str(self.pass_score) + " Shot : " + str(
+                self.shot_chance))
+            return True
+
+        # Decreasing and under cutoff
+        return False
+
+    def dribbler_has_ball(self):
+        return any(r.has_ball() for r in main.our_robots())
+
+    def on_enter_collecting(self):
+        self.remove_all_subbehaviors()
+
+        # 2 man to man defenders and 1 zone defender
+        defensive_forward = tactics.defensive_forward.DefensiveForward()
+        self.add_subbehavior(defensive_forward, 'defend', required=True)
+
+    def on_exit_collecting(self):
+        self.remove_all_subbehaviors()
+
+    def on_enter_dribbling(self):
+        self.dribbler = skills.dribble.Dribble()
+        self.dribble_start_pt = main.ball().pos
+
+        # Dribbles toward the best receive point
+
+        self.dribbler.pos, _ = evaluation.passing_positioning.eval_best_receive_point(
+            main.ball().pos,
+            main.our_robots(), AdaptiveAttack.MIN_PASS_DIST,
+            AdaptiveAttack.FIELD_POS_WEIGHTS,
+            AdaptiveAttack.NELDER_MEAD_ARGS,
+            AdaptiveAttack.DRIBBLING_WEIGHTS)
+
+        self.add_subbehavior(self.dribbler, 'dribble', required=True)
+
+        self.check_dribbling_timer = 0
+
+        if (not self.has_subbehavior_with_name('midfielders')):
+            self.midfielders = tactics.simple_zone_midfielder.SimpleZoneMidfielder(
+            )
+            self.add_subbehavior(
+                self.midfielders, 'midfielders', required=False, priority=10)
+
+    def execute_dribbling(self):
+        # Grab best pass
+        self.pass_target, self.pass_score = evaluation.passing_positioning.eval_best_receive_point(
+            main.ball().pos,
+            main.our_robots(), AdaptiveAttack.MIN_PASS_DIST,
+            AdaptiveAttack.FIELD_POS_WEIGHTS,
+            AdaptiveAttack.NELDER_MEAD_ARGS,
+            AdaptiveAttack.PASSING_WEIGHTS)
+
+        # Grab shot chance
+        self.shot_chance = evaluation.shooting.eval_shot(main.ball().pos)
+
+        # Recalculate dribbler pos
+        self.check_dribbling_timer += 1
+        if (self.check_dribbling_timer > self.check_dribbling_timer_cutoff):
+            self.check_dribbling_timer = 0
+            self.dribbler.pos, _ = evaluation.passing_positioning.eval_best_receive_point(
+                main.ball().pos,
+                main.our_robots(), AdaptiveAttack.MIN_PASS_DIST,
+                AdaptiveAttack.FIELD_POS_WEIGHTS,
+                AdaptiveAttack.NELDER_MEAD_ARGS,
+                AdaptiveAttack.DRIBBLING_WEIGHTS)
+
+        # TODO: Get list of top X pass positions and have robots in good positions to reach them
+        # Good positions can be definied by offensive / defensive costs
+        # Offensive positions move onto the ball in the direction of the goal
+        # Defensive cover the center of the field
+
```

Comment:
There are a few TODOs in these files are these yours, or are they just there from before? And if they're there from before do we still need them for these narrower plays?

Reply:
These are from when I initially created this. They're ideas to improve adaptive in the future. I think that only the last line 257 is changed with these narrower plays. Adaptive could still move all the offense robots to be even more agressive based on the pass position cost functions instead of just the basic positions